File: web/src/plugins/live-activity/native/LiveActivity.swift
// SPDX-License-Identifier: Apache-2.0
// LiveActivity plugin (ActivityKit). Drives a Live Activity from JS based on
// session busy/idle transitions. iOS 16.1+. No-op below 16.1.
import Foundation
import Capacitor
#if canImport(ActivityKit)
import ActivityKit
#endif

@objc(LiveActivity)
public class LiveActivity: CAPPlugin {
  // Type-erased storage: [sessionID: Activity<OpenCodeActivityAttributes>] held as Any,
  // because ActivityKit types require iOS 16.1+ and cannot annotate a stored property.
  private var activities: [String: Any] = [:]

  @objc func startActivity(_ call: CAPPluginCall) {
    guard let sessionID = call.getString("sessionID") else { call.reject("missing sessionID"); return }
    let title = call.getString("title") ?? "Session"
    let status = call.getString("status") ?? "busy"
    let detail = call.getString("detail") ?? "Working…"
    // progress: optional 0...1 fraction; omit/null → -1 (no progress bar).
    let progress: Double = call.getValue("progress") != nil ? (call.getDouble("progress") ?? -1) : -1
    #if canImport(ActivityKit)
    guard #available(iOS 16.1, *) else { call.resolve(); return }
    guard ActivityAuthorizationInfo().areActivitiesEnabled else { call.resolve(); return }
    if activities[sessionID] != nil { update(sessionID, title, status, detail, progress); call.resolve(); return }
    let attrs = OpenCodeActivityAttributes(sessionID: sessionID)
    let state = OpenCodeActivityAttributes.ContentState(title: title, status: status, detail: detail, progress: progress)
    do {
      let activity: Activity<OpenCodeActivityAttributes> = try Activity.request(attributes: attrs, contentState: state, pushType: nil)
      activities[sessionID] = activity
      call.resolve()
    } catch {
      call.reject("cannot start activity: \(error.localizedDescription)")
    }
    #else
    call.resolve()
    #endif
  }

  @objc func updateActivity(_ call: CAPPluginCall) {
    guard let sessionID = call.getString("sessionID") else { call.reject("missing sessionID"); return }
    let title = call.getString("title") ?? "Session"
    let status = call.getString("status") ?? "busy"
    let detail = call.getString("detail") ?? "Working…"
    let progress: Double = call.getValue("progress") != nil ? (call.getDouble("progress") ?? -1) : -1
    #if canImport(ActivityKit)
    guard #available(iOS 16.1, *) else { call.resolve(); return }
    update(sessionID, title, status, detail, progress)
    #endif
    call.resolve()
  }

  @objc func endActivity(_ call: CAPPluginCall) {
    guard let sessionID = call.getString("sessionID") else { call.reject("missing sessionID"); return }
    #if canImport(ActivityKit)
    guard #available(iOS 16.1, *) else { call.resolve(); return }
    if let raw = activities[sessionID] {
      let activity = raw as! Activity<OpenCodeActivityAttributes>
      Task {
        if #available(iOS 16.2, *) {
          await activity.end(nil, dismissalPolicy: .immediate)
        } else {
          await activity.end(using: nil, dismissalPolicy: .immediate)
        }
      }
      activities.removeValue(forKey: sessionID)
    }
    #endif
    call.resolve()
  }

  #if canImport(ActivityKit)
  @available(iOS 16.1, *)
  private func update(_ sessionID: String, _ title: String, _ status: String, _ detail: String, _ progress: Double) {
    guard let raw = activities[sessionID] else { return }
    let activity = raw as! Activity<OpenCodeActivityAttributes>
    let state = OpenCodeActivityAttributes.ContentState(title: title, status: status, detail: detail, progress: progress)
    Task { await activity.update(using: state) }
  }
  #endif
}

File: web/src/live-activity.ts
// SPDX-License-Identifier: Apache-2.0
// Based on opencode-remote-android by giuliastro (Apache-2.0). Modified for iOS, 2026.
// Live Activity driver: typed helpers around the LiveActivity Capacitor plugin,
// including progress computation from opencode todo lists. No-op off native platforms.
import { Capacitor } from "@capacitor/core"
import { LiveActivity } from "./plugins/live-activity"
import type { TodoItem } from "./types"

export type SessionStatus = "busy" | "retry" | "idle"

export interface LiveActivityState {
  sessionID: string
  title: string
  status: SessionStatus
  detail: string
  /** 0...1 fraction complete, or -1 to hide the progress bar. */
  progress: number
}

/** Hide the progress bar (no tour/todos data available). */
export const NO_PROGRESS = -1

/**
 * Compute a 0...1 progress fraction from an opencode todo list.
 * Todos whose status is "completed" count as done; any other status counts as
 * pending. Returns -1 when the list is empty/missing so the bar stays hidden.
 */
export function computeProgress(todos: TodoItem[] | null | undefined): number {
  if (!todos || todos.length === 0) return NO_PROGRESS
  const done = todos.filter((t) => (t.status ?? "").toLowerCase() === "completed").length
  if (done === 0) return 0
  if (done >= todos.length) return 1
  return done / todos.length
}

/** Build a short detail string from todos (most recent non-completed item). */
export function detailFromTodos(todos: TodoItem[] | null | undefined, fallback = "Working…"): string {
  if (!todos || todos.length === 0) return fallback
  const pending = todos.find((t) => (t.status ?? "").toLowerCase() !== "completed")
  const label = (pending ?? todos[todos.length - 1]).content
  if (!label) return fallback
  return label.length > 48 ? label.slice(0, 47) + "…" : label
}

function native(): boolean {
  return Capacitor.isNativePlatform()
}

/** Start (or update if already running) the Live Activity for a session. */
export function startActivity(state: LiveActivityState): Promise<void> {
  if (!native()) return Promise.resolve()
  return LiveActivity.startActivity({
    sessionID: state.sessionID,
    title: state.title,
    status: state.status,
    detail: state.detail,
    progress: state.progress,
  }).catch(() => undefined)
}

/** Update an already-running Live Activity (no-op if none exists for the session). */
export function updateActivity(state: LiveActivityState): Promise<void> {
  if (!native()) return Promise.resolve()
  return LiveActivity.updateActivity({
    sessionID: state.sessionID,
    title: state.title,
    status: state.status,
    detail: state.detail,
    progress: state.progress,
  }).catch(() => undefined)
}

/** End the Live Activity for a session. */
export function endActivity(sessionID: string): Promise<void> {
  if (!native()) return Promise.resolve()
  return LiveActivity.endActivity({ sessionID }).catch(() => undefined)
}
